### PR TITLE
[6.0] Depend on QueueManager contract

### PR DESCRIPTION
**Pull request by nmokkenstorm**
TLDR: This is a very minor upgrade, just making sure that `Illuminate\Queue\Worker` depends on the interface instead of the concrete class for `Illuminate\Queue\QueueManager`.

As for the use case, for a package I'm building [here](https://github.com/nmokkenstorm/laravel-queue-statistics) I wanted general access to the QueueManager singleton, and decorate it with some extra functionality, but the Worker would yell at me for injecting a different concrete class.

@X-Coder264 raised a valid point [here](https://github.com/laravel/framework/pull/29225) when I targeted this change on 5.8, and earlier [here](https://github.com/laravel/framework/pull/27826), so I implemented his proposed change and targeted master / 5.9 now.

As a side note, I tried a few things to make this work without having to extend the default QueueManager class and replace it so the package could be plug & play. In the end I found [this article](https://medium.com/@orobogenius/extending-core-laravel-bindings-97f409140fc3) which told me I could extend core bindings. I couldn't find this in the documentation, would a PR for adding that be appreciated?

### CODE REVIEW

**Comment by driesvints** [comment]
Heya, thanks for your pr. But I think we shouldn't inject the entire application into the worker so it becomes a hardcoded dependency. This way the component is hard-coupled to the foundation framework component. Maybe we should rethink how the `isDownForMaintenance` method is called instead.

**Comment by nmokkenstorm** [comment]
> Heya, thanks for your pr. But I think we shouldn't inject the entire application into the worker so it becomes a hardcoded dependency. This way the component is hard-coupled to the foundation framework component. Maybe we should rethink how the `isDownForMaintenance` method is called instead.

Hey @driesvints , thank you for reviewing.

The other solution I considered was adding `isDownForMaintenance` to the `Illuminate\Contracts\Queue\Factory` contract. To me this seems like a cleaner and more correct solution, mostly because that method isn't in either of the interfaces used by the QueueManager at the moment. I initially decided against it because adding a method feels like a change with more risk for breakage elsewhere in the (admittedly unlikely) event that someone is using that contract for something other than the QueueManager.

However, seeing as this PR will have to target 5.9 anyways, maybe adding the `isDownForMaintenance` method to the interface is the correct course of action here. I'd love to hear your thoughts.

**Comment by driesvints** [comment]
@nmokkenstorm That's not a perfect solution but probably a better one yeah. Let's go for that?

File: src/Illuminate/Queue/Worker.php
Comment (driesvints) [inline src/Illuminate/Queue/Worker.php]: Can you revert these changes?

File: src/Illuminate/Queue/Worker.php
Comment (driesvints) [inline src/Illuminate/Queue/Worker.php]: Revert this since the `isDownForMaintenance` is now on the manager contract.

File: src/Illuminate/Queue/QueueServiceProvider.php
Comment (driesvints) [inline src/Illuminate/Queue/QueueServiceProvider.php]: You can revert this now.

**Review by driesvints — CHANGES_REQUESTED** [review]

**Comment by driesvints** [comment]
Maybe also squash your commits.

**Comment by nmokkenstorm** [comment]
I've reverted unnecessary changes, and squashed the commits. This is the first time I'm doing the latter, so I hope it's correct, but it seems to be okay.

**Comment by X-Coder264** [comment]
My proposed solution there was just the quickest solution, not the way I'd go about implementing this.

IMO, the manager `isDownForMaintenance` is called only in the worker and I don't see the need to have that method there (and thus no value to add it to the contract).

The worker can receive a `callable|bool` in its constructor which can be a closure passed from the service provider doing `$this->app->isDownForMaintenance()`.

That way the worker does not need to depend on the entire application/container and that method doesn't have to be added to the contract and it can be deleted from the queue manager class.

**Comment by driesvints** [comment]
@X-Coder264 that's a great idea 👍 

@nmokkenstorm do you agree with this? If so you can make these changes if you like.

**Comment by nmokkenstorm** [comment]
I had considered passing a boolean, but consider the following scenario:

- One or more workers get started while the application is not in maintenance mode, get passed the proposed boolean flag.
- Maintenance mode gets activated.
- Workers will never pick up the fact that maintenance mode is activated.

I'm not too sure what the conventions are for putting apps in maintenance mode, if people use `php artisan queue:restart` afterwards the above is not an issue. We do risk creating a hard to track down  bug for people relying on the maintenance mode check being 'live'.

Passing a closure in QueueServiceProvider might work:

```
protected function registerWorker()
{
    $this->app->singleton('queue.worker', function () {
        
        $maintenanceModeCheck = function () use ($app) {
            return $app->isDownForMaintenance();
        }

        return new Worker(
            $this->app['queue'], $this->app['events'], $this->app[ExceptionHandler::class], $maintenanceModeCheck
        );
    });
}
```

Optionally the closure could be inlined in the Worker constructor, but I'm not sure how I feel about nesting things that deep. I'll happily make the changes, if you agree with the above.

**Comment by X-Coder264** [comment]
Yes, that's way I said that it needs to be a closure passed from the service provider :)

**Comment by nmokkenstorm** [comment]
> Yes, that's way I said that it needs to be a closure passed from the service provider :)

My apologies, I might've misunderstood your comment about the paramater being a `callable|bool`

**Comment by X-Coder264** [comment]
Np, I said `callable|bool` so that someone can pass directly a bool value if he wants to do so, but the default implementation provided by the framework should pass the closure.

**Comment by driesvints** [comment]
@X-Coder264 hmm, that might be a little too unconventional no? Maybe we should just rename the method `isDownForMaintenance` method on the `Factory` interface to something more generic that makes sense? 

What I'm also asking myself: why can't the queue continue the run when the app is in maintenance mode? In my view maintenance mode was always something front-end wise to inform the user that they can't interact with an app for a while. But that doesn't seems applicable to a queue imo.

**Comment by nmokkenstorm** [comment]
> @X-Coder264 hmm, that might be a little too unconventional no? Maybe we should just rename the method `isDownForMaintenance` method on the `Factory` interface to something more generic that makes sense?
> 
> What I'm also asking myself: why can't the queue continue the run when the app is in maintenance mode? In my view maintenance mode was always something front-end wise to inform the user that they can't interact with an app for a while. But that doesn't seems applicable to a queue imo.

@driesvints the framework specifically provides functionality for Jobs running / not running in maintenance mode as defined [here](https://laravel.com/docs/5.0/queues). 5.8 documentation doesn't actually state this, but the code does reflect it.

**Comment by driesvints** [comment]
@nmokkenstorm gotcha. Then maybe @X-Coder264's suggestion is the better approach.

**Comment by nmokkenstorm** [comment]
> Np, I said `callable|bool` so that someone can pass directly a bool value if he wants to do so, but the default implementation provided by the framework should pass the closure.

I think passing a boolean here is never a good idea for the reasons I outline above. Honestly I don't expect users to manually create workers anyways, and in the event that they do and they _really_ want a constant expression, a direct return from the callable would fit their needs.

**Comment by driesvints** [comment]
Hmm, after seeing these change I'm beginning to suspect that the "cleaner" approach is just adding the method on the interface after all. The app is already a big dependency on the factory. I don't think it matters much if the Factory has the maintenance check on its contract. 

@X-Coder264 @nmokkenstorm what do you think?

**Comment by nmokkenstorm** [comment]
The new functionality is in, I also took the liberty of adding a test because apparently this wasn't covered yet.

The one thing I am not sure about is how to call the closure. Right now I do it by function invocation but maybe `call_user_func` is better here. I'm open for suggestions.

**Comment by nmokkenstorm** [comment]
> Hmm, after seeing these change I'm beginning to suspect that the "cleaner" approach is just adding the method on the interface after all. The app is already a big dependency on the factory. I don't think it matters much if the Factory has the maintenance check on its contract.
> 
> @X-Coder264 @nmokkenstorm what do you think?

The Worker depends on the Manager, which depends on Application, so you're correct in that de dependency tree doesn't change that much. The way @X-Coder264 suggested it is definitely 'safer' with regards to breaking backwards compatibility.

I did do a quick search on public Github repositories to check if anyone is using the contract outside of the manager, but that didn't work as well as I would've liked it to.

**Comment by X-Coder264** [comment]
@driesvints  The worker depends now on the factory contract. What are the dependencies of the factory contract implementation are irrelevant for the worker. And about the factory/QueueManager concrete implementation, it can now be also decoupled from the foundation application. Since the `isDownForMaintenance` method is being removed from it in this PR, the only two dependencies that it has now are the config and event dispatcher contracts which can be directly injected instead of the whole container/application and thus the queue component in general will be a lot less coupled to the framework foundation.

**Comment by nmokkenstorm** [comment]
> ... the only two dependencies that it has now are the config and event dispatcher contracts which can be directly injected instead of the whole container/application and thus the queue component in general will be a lot less coupled to the framework foundation.

Huh, I hadn't even considered that. I'm not sure if this PR is the place to change that, but I might as well do that if I'm doing work there anyways. @driesvints  @taylorotwell , thoughts?

**Comment by driesvints** [comment]
Let's just see what Taylor says.

**Comment by nmokkenstorm** [comment]
To summarise after the weekend:

The goal of this PR is to make `Illuminate\Queue\Worker` not depend on a concrete instance of `Illuminate\Queue\Manager` but on the contract offered in `Illuminate\Contracts\Queue\Factory`.

The initial approach was to inject `Illuminate\Foundation\Application` into the worker, which @driesvints thought was a bit overkill.

The other two possible approaches here are adding the `isDownForMaintenance` method to the contract or adding a callable as a fourth argument to the constructor of `Illuminate\Queue\Worker`. 

The latter option is the current state of the PR, but I feel that that injecting the Application into the Worker is still the cleanest way to go about this, minimal impact with regards to breaking changes.

**Comment by nmokkenstorm** [comment]
Related to the above, [this](https://github.com/laravel/framework/commit/9227682233e09dc40450aea7bbb481de83a18bda) commit is now conflicting in the tests, but that should be an easy rebase fix.

**Comment by X-Coder264** [comment]
> The latter option is the current state of the PR, but I feel that that injecting the Application into the Worker is still the cleanest way to go about this, minimal impact with regards to breaking changes.

I disagree, there should be no reason that the worker should depend on the entire application (I agree with @driesvints that it's overkill). As far as breaking changes go it's the same impact for both approaches as a new argument has to be added to the constructor either way.

TL;DR; I don't see any reason why the former option would be chosen over the latter (the current state of the PR)..

**Comment by nmokkenstorm** [comment]
@X-Coder264  you're correct, either of the three options work, which is why we want to defer to @taylorotwell here

**Comment by franzliedke** [comment]
@nmokkenstorm @X-Coder264 Thank you both so much! This makes using the Queue component by itself significantly more easy.

Now I just need to wait for the release. :hourglass_flowing_sand:

**Comment by franzliedke** [comment]
Should the `setManager()` function (which isn't used within the framework) be renamed to `setFactory()` as well? (Aliasing the `Factory` contract to `QueueManager` is quite confusing, hehe...)

**Pull request merged**
